src/net/poll.rs

```diff
@@ -159,32 +159,51 @@ impl NetworkState {
             server_event: mio::Token(next_server_event),
         };
 
-        assert!(!self.event_map.contains_key(&next_server_event));
+        assert!(!self.event_map.contains_key(&next_server_event), "BUG: failed to generate an unused server event ID");
 
         self.servers.push(network_server);
         self.event_map.insert(next_server_event, 0);        // server events always mapped to 0
 
         Ok(next_server_event)
     }
 
-    /// Register a socket for read/write notifications with this poller
-    pub fn register(&mut self, server_event_id: usize, event_id: usize, sock: &mio_net::TcpStream) -> Result<(), net_error> {
+    /// Register a socket for read/write notifications with this poller.
+    /// Try to use the given hint_event_id value, but generate a different event ID if it's been
+    /// taken.
+    /// Return the actual event ID used (it may be different than hint_event_id)
+    pub fn register(&mut self, server_event_id: usize, hint_event_id: usize, sock: &mio_net::TcpStream) -> Result<usize, net_error> {
+        let hint_event_id = hint_event_id % self.event_capacity;
+        if let Some(x) = self.event_map.get(&server_event_id) {
+            if x != &0 {
+                // not a server event
+                error!("Server event ID {} not mapped to a server token, but to {}", &server_event_id, x);
+                return Err(net_error::RegisterError);
+            }
+        }
+        else {
+            // not a server event
+            error!("Not a server event ID: {}", &server_event_id);
```

Comment:
What could have led to this error? Should this panic?

Reply:
This can only happen due to a bug elsewhere -- i.e. you tried to register a socket on a server event that doesn't exist (and there can only be two -- the p2p network server and the http server).  Can make this a panic.

Reply:
Yep -- let's switch to panic, since it'd indicate a bug.

Reply:
Fixed in 986e210